tokio/src/io/unix/pipe.rs

```diff
@@ -0,0 +1,1208 @@
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_receiver<P: AsRef<Path>>(&self, path: P) -> io::Result<Receiver> {
+        let file = self.open(path.as_ref(), PipeEnd::Receiver)?;
+        Receiver::from_file_unchecked(file)
+    }
+
+    /// Creates a [`Sender`] from a FIFO file with the options specified by `self`.
+    ///
+    /// This function will open the FIFO file at the specified path, possibly
+    /// check if it is a pipe, and associate the pipe with the default event
+    /// loop for writing.
+    ///
+    /// # Errors
+    ///
+    /// If the file type check fails, this function will fail with `io::ErrorKind::InvalidInput`.
+    /// If the file is not opened in read-write access mode and the file is not
+    /// currently open for reading, this function will fail with `ENXIO`.
+    /// This function may also fail with other standard OS errors.
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_sender<P: AsRef<Path>>(&self, path: P) -> io::Result<Sender> {
+        let file = self.open(path.as_ref(), PipeEnd::Sender)?;
+        Sender::from_file_unchecked(file)
+    }
+
+    fn open(&self, path: &Path, pipe_end: PipeEnd) -> io::Result<File> {
+        let mut options = std::fs::OpenOptions::new();
+        options
+            .read(pipe_end == PipeEnd::Receiver)
+            .write(pipe_end == PipeEnd::Sender)
+            .custom_flags(libc::O_NONBLOCK);
+
+        #[cfg(target_os = "linux")]
+        if self.read_write {
+            options.read(true).write(true);
+        }
+
+        let file = options.open(path)?;
+
+        if !self.unchecked && !is_fifo(&file)? {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput, "not a pipe"));
+        }
+
+        Ok(file)
+    }
+}
+
+impl Default for OpenOptions {
+    fn default() -> OpenOptions {
+        OpenOptions::new()
+    }
+}
+
+#[derive(Clone, Copy, PartialEq, Eq, Debug)]
+enum PipeEnd {
+    Sender,
+    Receiver,
+}
+
+/// Writing end of a Unix pipe.
+///
+/// It can be constructed from a FIFO file with [`OpenOptions::open_sender`].
+///
+/// Opening a named pipe for writing involves a few steps.
+/// Call to [`OpenOptions::open_sender`] might fail with an error indicating
+/// different things:
+///
+/// * [`io::ErrorKind::InvalidInput`] - There is no FIFO file at specified path.
```

Comment:
If the file doesn't exist at all, then the error we get is something else.

Reply:
Updated the docs to distinguish between these two error kinds.